### PR TITLE
Use Ranch for TCP acceptor pool

### PR DESCRIPTION
**Pull request by essen**
Part of https://github.com/rabbitmq/rabbitmq-server/issues/260.

To be merged alongside:

* https://github.com/rabbitmq/rabbitmq-common/pull/8
* https://github.com/rabbitmq/rabbitmq-erlang-client/pull/22
* https://github.com/rabbitmq/rabbitmq-stomp/pull/46
* https://github.com/rabbitmq/rabbitmq-mqtt/pull/42

### CODE REVIEW

**Comment by michaelklishin** [comment]
Depending on a future Ranch version is fine now with erlang.mk as long as we are confident in its stability.

**Comment by essen** [comment]
I want to make sure all tests pass, then make Ranch 1.2 official, before 3.6.0 is released (will be just renaming the version if I did my job properly). Right now it's depending on 1.2.0-rc.1.

**Comment by essen** [comment]
All (working) Erlang tests pass on my machine.

The Java client test suite also passes on my machine: https://gist.github.com/essen/b9b4cc5e9d0a3b433833

Will see if there's more interesting things to test (maybe inside plugins) and finalize the PRs.

**Comment by michaelklishin** [comment]
Broker cannot start with all defaults for me (on OS X):

```
BOOT FAILED
===========

Error description:
   {could_not_start,rabbit,
    {{case_clause,
      {error,
       {{shutdown,
         {failed_to_start_child,
          {ranch_listener_sup,{acceptor,{0,0,0,0,0,0,0,0},5672}},
          {shutdown,
           {failed_to_start_child,ranch_acceptors_sup,
            {function_clause,
             [{ranch,filter_user_options,
               [[inet6,binary,
                 {backlog,128},
                 {nodelay,true},
                 {linger,{true,0}},
                 {exit_on_close,false},
                 {active,false},
                 {packet,raw},
                 {reuseaddr,true}],
                [backlog,ip,linger,nodelay,port,raw,send_timeout,
                 send_timeout_close]],
               [{file,"src/ranch.erl"},{line,127}]},
              {ranch,filter_user_options,2,
               [{file,"src/ranch.erl"},{line,129}]},
              {ranch,filter_user_options,2,
               [{file,"src/ranch.erl"},{line,129}]},
              {ranch,filter_options,3,[{file,"src/ranch.erl"},{line,124}]},
              {ranch_tcp,listen,1,[{file,"src/ranch_tcp.erl"},{line,65}]},
              {ranch_acceptors_sup,init,1,
               [{file,"src/ranch_acceptors_sup.erl"},{line,30}]},
              {supervisor,init,1,[{file,"supervisor.erl"},{line,272}]},
              {gen_server,init_it,6,
               [{file,"gen_server.erl"},{line,328}]}]}}}}},
        {child,undefined,'rabbit_tcp_listener_sup_:::5672',
         {tcp_listener_sup,start_link,
          [{0,0,0,0,0,0,0,0},
           5672,ranch_tcp,
           [inet6,binary,
            {backlog,128},
            {nodelay,true},
            {linger,{true,0}},
            {exit_on_close,false},
            {active,false},
            {packet,raw},
            {reuseaddr,true}],
           rabbit_connection_sup,[],
           {rabbit_networking,tcp_listener_started,[amqp]},
           {rabbit_networking,tcp_listener_stopped,[amqp]},
           "TCP Listener"]},
         transient,infinity,supervisor,
         [tcp_listener_sup]}}}},
     [{rabbit_networking,start_listener0,4,
       [{file,"src/rabbit_networking.erl"},{line,305}]},
      {rabbit_networking,'-start_listener/4-lc$^0/1-0-',4,
       [{file,"src/rabbit_networking.erl"},{line,294}]},
      {rabbit_networking,start_listener,4,
       [{file,"src/rabbit_networking.erl"},{line,294}]},
      {rabbit_networking,'-boot_tcp/0-lc$^0/1-0-',1,
       [{file,"src/rabbit_networking.erl"},{line,125}]},
      {rabbit_networking,boot_tcp,0,
       [{file,"src/rabbit_networking.erl"},{line,125}]},
      {rabbit_networking,boot,0,
       [{file,"src/rabbit_networking.erl"},{line,120}]},
      {rabbit_boot_steps,'-run_step/2-lc$^1/1-1-',1,
       [{file,"src/rabbit_boot_steps.erl"},{line,49}]},
      {rabbit_boot_steps,run_step,2,
       [{file,"src/rabbit_boot_steps.erl"},{line,49}]}]}}
```

Nothing stands on the list of socket options.

**Comment by essen** [comment]
Probably binary.

Though it's an issue that I fixed in Ranch 1.2.0-rc.1 so check that you got the right version pulled in.

**Comment by michaelklishin** [comment]
`make clean` in the server, common and umbrella did not help but manually removing `deps/ranch` did. Things seem to work now.

**Comment by michaelklishin** [comment]
Java client, Bunny, March Hare, and Erlang client test suites pass on multiple OS X and Linux hosts.

**Comment by essen** [comment]
Good to know thanks. Will continue on Tuesday. If MQTT is really all that's left, I'll most likely also be done on the same day.

**Comment by essen** [comment]
MQTT is done. I will update Ranch to 1.2.0, review the PRs one more time and we should be good.

**Comment by essen** [comment]
OK. Please review!

**Comment by michaelklishin** [comment]
This seems to be ready for merging, we now need to wait for some build system updates to accommodate dependencies more than 1 level deep (rabbit => rabbit_common => ranch).

**Comment by michaelklishin** [comment]
Testing GH notifications.

**Comment by michaelklishin** [comment]
Ditto.

**Comment by michaelklishin** [comment]
@essen I'm observing an exception that only this branch has in the SASL log:

```
=CRASH REPORT==== 19-Nov-2015::17:25:04 ===
  crasher:
    initial call: rabbit_reader:init/4
    pid: <0.6803.0>
    registered_name: []
    exception exit: {noproc,{gen_server,call,[undefined,delete_all,infinity]}}
      in function  gen_server:call/3 (gen_server.erl, line 212)
      in call from rabbit_reader:close_connection/1 (src/rabbit_reader.erl, line 687)
      in call from rabbit_reader:send_error_on_channel0_and_close/4 (src/rabbit_reader.erl, line 1463)
      in call from rabbit_reader:handle_input/3 (src/rabbit_reader.erl, line 998)
      in call from rabbit_reader:recvloop/4 (src/rabbit_reader.erl, line 444)
      in call from rabbit_reader:run/1 (src/rabbit_reader.erl, line 426)
      in call from rabbit_reader:start_connection/4 (src/rabbit_reader.erl, line 384)
    ancestors: [<0.6801.0>,<0.280.0>,<0.279.0>,<0.278.0>,rabbit_sup,
                  <0.169.0>]
    messages: [{'EXIT',#Port<0.19890>,normal}]
    links: [<0.6801.0>]
    dictionary: [{process_name,
                      {rabbit_reader,
                          <<"[FE80::1]:52718 -> [FE80::1]:5672">>}}]
    trap_exit: true
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 4046
  neighbours:

=SUPERVISOR REPORT==== 19-Nov-2015::17:25:04 ===
     Supervisor: {<0.6801.0>,rabbit_connection_sup}
     Context:    child_terminated
     Reason:     {noproc,{gen_server,call,[undefined,delete_all,infinity]}}
     Offender:   [{pid,<0.6803.0>},
                  {name,reader},
                  {mfargs,
                      {rabbit_reader,start_link,
                          [<0.6802.0>,
                           {acceptor,{0,0,0,0,0,0,0,0},5672},
                           #Port<0.19890>]}},
                  {restart_type,intrinsic},
                  {shutdown,4294967295},
                  {child_type,worker}]
```

in other words, when connection process tries to close a client connection, `queue_collector`, which tracks and cleans up exclusive queues, is `undefined`. It can only be undefined before `connection.tune*` steps complete. I do not observe this with master.

The easiest way to reproduce is with [Bunny](https://github.com/ruby-amqp/bunny):

```
# start rabbitmq
cd /path/to/deps/rabbit
make run-broker PLUGINS='rabbitmq_management rabbitmq_consistent_hash_exchange'

# with Ruby 2.0 or later
git clone https://github.com/ruby-amqp/bunny
cd bunny
gem install bundler
bundle install
export RABBITMQCTL=/path/to/deps/rabbit/scripts/rabbitmqctl
./bin/ci/before_build
rspec -cfd spec/higher_level_api/integration/connection_recovery_spec.rb
```

and then tail SASL logs.

**Comment by michaelklishin** [comment]
I did a rebuild while switching to master and back, and no longer have this exception. Asking more people to give this a try.

**Comment by michaelklishin** [comment]
@essen nevermind the above. The issue is unrelated to Ranch but rather my error message refactoring, which now leads all early terminating connections (e.g. due to authentication issues) through a function that tries to clean up exclusive queues. Sometimes there would be nothing to clean up, so it needs to be handled.

File: src/rabbit_vm.erl [hunk 1]
```diff
@@ -134,7 +134,11 @@ interesting_sups0() ->
     PluginProcs   = plugin_sups(),
     [MsgIndexProcs, MgmtDbProcs, PluginProcs].
 
-conn_sups()     -> [rabbit_tcp_client_sup, ssl_connection_sup, amqp_sup].
+%% @todo I have doubts about this ssl_connection_sup and the
```
Comment (michaelklishin) [inline src/rabbit_vm.erl]: @essen is this still relevant? `amqp_sup` is used in the Erlang client.

Comment (essen) [inline src/rabbit_vm.erl]: I will send a PR for that also.

**Pull request merged**